### PR TITLE
 Scan Policy Manager Policy multiplies while modifying the name

### LINKED ISSUE #7134: Scan Policy Manager Policy multiplies while modifying the name
### Describe the bug

When you change the Policy Name more than two Time(e.g. from upper Case to Lower-Case) then the item got duplicated, after removing one item the other cannot be modified anymore but removed.

### Steps to reproduce the behavior

1. Start Zap
2. Open Scan Policy Manager (CTRL-P)/ Analyse -> Scan Policy Manager
3. Add a new Policy
4. Name it "scan"
5. Press Ok
6. Modify the "scan"-item
7. Change Policy-Name to "Scan" and press okay again
8. Modify the item again (notice that the Policy-Name is already "scan" instead of "Scan") change the name again to "sCan" and press okay
9. Repeating step 8 to different Names creates even more Items and all of them seems to be the one which we have created first.
10. Remove one item and you cannot modify the other ones.

Attached a video how to reproduce the bug
[2022-03-10 09-59-46.zip](https://github.com/zaproxy/zaproxy/files/8221998/2022-03-10.09-59-46.zip)

### Expected behavior

The policy name is changed and no other item is created.

### Software versions

Version 2.11.1

### Screenshots

zap_policy_name_bug

### Errors from the zap.log file

_No response_

### Additional context

_No response_

### Would you like to help fix this issue?

- [X] Yes

### PR DESCRIPTION
**Pull request by marvUta**
Added condition to delete old policy file in case of a namechange, plus some tests to check it. Testfile got another file.exists() method otherwise the usuall exists()-method wouldn't differ between "Test" and "test". I've tested it on a Windows-System.

Fix  #7134.

### CODE REVIEW

**Comment by thc202** [comment]
Some checks are failing.

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

The code need a spotlessApply: `./gradlew spotlessApply`

**Comment by marvUta** [comment]
Thanks for the advice, spotlessApply fixed it and all checks passed on my side. I've signed off the commit and pushed one more time. 
As long the analysis were starting immediately after my push I assume I don't need to merge or update the request, right?

**Comment by kingthorin** [comment]
All commits need signoff. You may as well squash or fixup as part of an interactive rebase and just push one commit.

**Comment by marvUta** [comment]
All right, do I need a rebase or does another commit with "git commit --amend --signoff" do it as well?

**Comment by kingthorin** [comment]
No amend only applies to the latest commit. If you squash them then do the amend that’ll work.

**Comment by kingthorin** [comment]
I’ll put together some instructions. I don’t have power this morning, replying here on my cell phone. Will try to get the details once power is back.

**Comment by marvUta** [comment]
Okay thanks, somehow I pulled the unsigned push in again after I've done the rebase. At the end I've done the following.

1. rebase one before the sign off where missing
2. force push instead of a pull

**Comment by kingthorin** [comment]
Sorry for the delay, I’m still without power.

**Comment by kingthorin** [comment]
Here's the instructions, if this seems too daunting or you get into problems let me know and I can take care of it.

```bash
git rebase upstream/main
// You will probably get a warning here about skipped commits, just ignore it

git rebase --interactive 879792acb5a7cf381b6698c721e9dee055eed407
// An editor will open at this point, set things as follows
// By defualt it's vim so use "i" to insert/edit then "esc" to exit insert/edit mode and ":wq <enter>" to write and quit

d 2719c5e16 added condition to delete old policy file in case of a namechange added some testcases(policy namechages) for policymanager
pick c909e4935 delete old policy-file in case of a name change added testcases
f c491d882e added condition to delete old policy file in case of a namechange added some testcases(policy namechages) for policymanager
d cebd2cef6 delete old policy-file in case of a name change added testcases

# Rebase 879792acb..77d885f54 onto 879792acb (4 commands)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
#                    commit's log message, unless -C is used, in which case
#                    keep only this commit's message; -c is same as -C but
#                    opens the editor
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
# .       create a merge commit using the original merge commit's
# .       message (or the oneline, if no original merge commit was
# .       specified); use -c <commit> to reword the commit message
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
~

$ git log -2
// Should look like this

commit 61837ed8db1f58e62da91c9f8c0a14cf99661156 (HEAD -> main-marv)
Author: Marvin Uta <mux.leet@gmail.com>
Date:   Sun May 22 10:56:42 2022 +0200

    added condition to delete old policy file in case of a namechange
    added some testcases(policy namechages) for policymanager

    Signed-off-by: Marvin Uta <mux.leet@gmail.com>

commit 879792acb5a7cf381b6698c721e9dee055eed407 (tag: w2022-05-23, upstream/main, origin/main, origin/HEAD, main)
Merge: 3b76bea36 0d749443d
Author: Rick M <kingthorin@users.noreply.github.com>
Date:   Mon May 23 07:37:00 2022 -0400

    Merge pull request #7214 from DavidCervigni/patch-1

    Enable debug option for Mac in zap.sh

$ git commit --amend --signoff
// You'll be dropped into an editor again, set things like this:
Scan Policy Manager Policy multiplies while modifying the name

Delete old policy-file in case of a name change
and added some testcases(policy namechages) for policymanager

Signed-off-by: Marvin Uta <mux.leet@gmail.com>

$ git push --force
```

**Comment by kingthorin** [comment]
Okay, I'll tiddy it up. Once I've done that it'll just take a line or two for you to sync your local with your fork so that you're in the same state as the PR once I've fixed it up.

**Comment by marvUta** [comment]
Okay awesome. If we meet somehow drinks are on me :D

**Comment by kingthorin** [comment]
@marvUta please make sure that looks right now.

Before you do anything further please also:

```bash
git fetch origin
git reset --hard origin/main
```

(That assumes origin is set to your forked repo, ex: https://github.com/marvUta/zaproxy.git)

**Comment by marvUta** [comment]
Okay I've done a fetch upstream on github and git fetch origin(cli) after that "git reset --hard origin/main" respondig with the following:
`HEAD is now at 25aa88711 Merge branch 'zaproxy:main' into main`

**Comment by kingthorin** [comment]
> Okay I've done a fetch upstream on github

That wasn't necessary, and actually means there's an extra commit here again.

Glad the rest worked out.

**Comment by kingthorin** [comment]
@marvUta let me know if you need me to tackle it?

**Comment by kingthorin** [comment]
The file permissions he was referring to are those in the PR. Ex 644 being changed to 755:
image

I guess the added file (unittest) is also 755, but it doesn't show in the GitHub web UI because the file is just being added.

(At least that's my understanding.)

**Comment by marvUta** [comment]
Ah I got it. Yeah I'll do it tomorrow.

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (thc202) [inline zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java:106]: `static`

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyManager.java
Comment (marvUta) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyManager.java]: I've harden the if-condition and removed the comment

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (marvUta) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: Updaed to policymanger

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (marvUta) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: changed all tests to should

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (marvUta) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: formatted the code

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (marvUta) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: maked the helperfunction static

File: zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java
Comment (kingthorin) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: It’s not just “should” it needs to be proper English and describe what’s being tested.

Comment (marvUta) [reply zap/src/test/java/org/zaproxy/zap/extension/ascan/PolicyManagerUnitTest.java]: Okay my bad I just skipped the details. Thanks for being so patient. I'll rename today and after fixing the other ones.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyManager.java
Comment (marvUta) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyManager.java]: The problem for new policies still exists and I'll add a check with existings policies today

**Comment by marvUta** [comment]
Thanks for noticing it, I've seen the Java CI/builds failed again because of this.

I was thinking about the issue on case insensitive systems if you add two policies with the same name(e.g. Test and test) and I would recommend to add a check while pressing the OK button. In that case the use can be informed that the name is already taken. Otherwise if the policy-file isn't saved without any notification it can be missleading.

**Comment by kingthorin** [comment]
> I was thinking about the issue on case insensitive systems if you add two policies with the same name(e.g. Test and test) and I would recommend to add a check while pressing the OK button. In that case the use can be informed that the name is already taken. Otherwise if the policy-file isn't saved without any notification it can be missleading.

That seems reasonable to me. Please go ahead with that and include it as part of this PR. (@psiinon @thc202 do you agree?)

**Comment by thc202** [comment]
Yes.

**Comment by ricekot** [comment]
Alternatively we could make the behaviour uniform across all systems by making the scan policy names case insensitive within ZAP? But doing so might break existing workflows for some people.

**Comment by thc202** [comment]
That's what needs to be done to fix the issue.

**Comment by marvUta** [comment]
Sorry for the late response, I was some days off. Actually I've already implemented it at the point where the User presss the OK-Button if a name was change/created a new policy. For exiting Policies with Upper-LowerCase difference the User will be informed until a name change not when the policy is saved. 
Shall this be extended to the point where the policy file is saved as well?

Out of curiousity is it possible to create/modify/delete scan policies via scripts which may bypass the check?

**Comment by thc202** [comment]
Rebased to pick the latest checks, I have only one comment will add it now and address it shortly.

> Shall this be extended to the point where the policy file is saved as well?

I think it's fine as is.

> Out of curiosity is it possible to create/modify/delete scan policies via scripts which may bypass the check?

Yes, that's possible (the user can even do that outside ZAP by manually creating/changing/deleting the files).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyAllCategoryPanel.java
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyAllCategoryPanel.java:536]: This should be changed to be case insensitive as well, otherwise it prevents changing the case of the policy itself.

Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/PolicyAllCategoryPanel.java:536]: Or not, this would require further validation if a policy with same case already exists in case sensitive file systems.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Comment by thc202** [comment]
For the record, while https://github.com/zaproxy/zaproxy/pull/7295#issuecomment-1149635805 is the best solution going forward that would require more changes and better have this issue fixed in next release.

**Comment by thc202** [comment]
@marvUta how would you like to be credited (e.g. name, handle)?
https://www.zaproxy.org/docs/desktop/credits/#zap-extended-team

**Comment by marvUta** [comment]
> @marvUta how would you like to be credited (e.g. name, handle)? https://www.zaproxy.org/docs/desktop/credits/#zap-extended-team

@thc202 Thanks for the credits, I am feeling honored. Simply "Marvin Uta"

**Pull request merged**
